stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,82 @@
+#! input:  [v, ...]
+#! output: [v >> 48, ...]
+proc.shr_48
+    u32split
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 32, ...]
+proc.shr_32
+    u32split
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 16, ...]
+proc.shr_16
+    u32split
+    push.65536
+    mul
+    u32split
+    push.4294967296
+    mul
+    swap.2
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    add
+    add
+end
+
+#! Fetches the keyed value of a Sparse Merkle tree with the provided root.
+#! cycles: 91
+#! input:  [K, R, ...]
+#! output: [d, i, V, K, R, ...]
+export.get
```

Comment:
1. I thought of preserving that as a helper for the user, but will remove it as he indeed might not need it.

2. The main goal of `smt::get` is to be performed under `100` cycles, so it has no tolerance for edge cases. With it, we introduced a couple of constraints

- We expect `V` to be in the advice map (hence, the key must exist in the set)
- We check the opening of `H_d(K, V) -> R` via `mtree_get`, so we verify the output of the advice injector

Also, we must distinct what is a leaf value, and what is a node value. The leaf value is the one in the advice map, represented by `K |-> V`, and the node value is `H_{depth}(K, V)`. The zeroed element is the `node value`; zeroing the `leaf value` would invalidate a key that maps to zero (that seems to be a plausible case).

We currently don't return the `node value`, but it is available on the advice stack, so we can easily include it. It would change the current routine to be as

```
#! input:  [K, R, ...]
#! output: [d, i, N, V, K, R, ...]
```

In order to support empty keys, I propose we create a more robust variant `smt::get_option` that will check the node value, and assert the opening via `mtree_get` using the empty constant. This, however, will be more expensive than `smt::get`

3. Fixed

Reply:
> The main goal of `smt::get` is to be performed under `100` cycles, so it has no tolerance for edge cases.

I would say that the main goal is to implement the expected interface (doing it under 100 cycles is another goal, but it shouldn't come at the expense of the first one). As is, the `smt::get` method cannot be invoked by the user because the user would not know whether a given key is in the SMT or not. So, we should probably stick to the original spec.

> The leaf value is the one in the advice map, represented by `K |-> V`

As I mentioned in one of my other comments, this will not be the case. We cannot maintain this mapping in the advice map as it may change during the runtime of the program.

Reply:
> It seems to me the most common use case will be to just fetch the value, if existing, via `smt::get`. We should provide other functions that will work as simple - and optimal - as possible

The user code cannot make assumptions about whether the value exists in the SMT or not. The core property of the SMT is that it returns a value for any key without failing - otherwise, a different data structure can be used.

Reply:
Makes sense!

To summarize, the advice injector will

$$S_{operand}: [K, R, ...] \mapsto S_{advice}: [d, V]$$

And `smt::get` will

$${S_{operand}: [K, R, ...] \mapsto S_{operand}: [d, i, V, N ,R, ...]$$

We will have all `(d, i, V, N)` available on MASM, we can just drop them or expose to the user

Reply:
> We will have all `(d, i, V, N)` available on MASM, we can just drop them or expose to the user. I think it's better if we expose all of them as the user might drop the values himself, if he doesn't need them.

I don't think we should expose them. These are internal implementation details and, in theory, could change in the future. As far as the user is concerned, SMT acts as a key-value map and the stack transition is:
```
[K, R, ...] -> [V, R, ...]
```

Reply:
Updated to

$$`adv.smtget` | S_{operand}: [K, R, ...] \mapsto S_{advice}: [d, f, V?]$$
$$`smt::get` | S_{operand}: [K, R, ...] \mapsto S_{operand}: [V, F]$$